packages/eslint-plugin/docs/rules/padding-line-between-statements.md

```diff
@@ -0,0 +1,79 @@
+        { "blankLine": LINEBREAK_TYPE, "prev": MODIFIER_TYPE STATEMENT_TYPE, "next": MODIFIER_TYPE STATEMENT_TYPE },
+        { "blankLine": LINEBREAK_TYPE, "prev": MODIFIER_TYPE STATEMENT_TYPE, "next": MODIFIER_TYPE STATEMENT_TYPE },
+        ...
+    ]
+}
+```
+
+- `LINEBREAK_TYPE` is one of the following.
+
+  - `"any"` just ignores the statement pair.
+  - `"never"` disallows blank lines.
+  - `"always"` requires one or more blank lines. Note it does not count lines that comments exist as blank lines.
+
+- `STATEMENT_TYPE` is one of the following, or an array of the following.
+
+  - `"*"` is wildcard. This matches any statements.
+  - `"block"` is lonely blocks.
+  - `"block-like"` is block like statements. This matches statements that the last token is the closing brace of blocks; e.g. `{ }`, `if (a) { }`, and `while (a) { }`. Also matches immediately invoked function expression statements.
+  - `"break"` is `break` statements.
+  - `"case"` is `case` clauses in `switch` statements.
+  - `"exports"` is `exports` statements of CommonJS; e.g. `module.exports = 0`, `module.exports.foo = 1`, and `exports.foo = 2`. This is a special case of assignment.
+  - `"require"` is `require` statements of CommonJS; e.g. `const foo = require("foo")`. This is a special case of variable declarations.
+  - `"class"` is `class` declarations.
+  - `"const"` is `const` variable declarations, both single-line and multiline.
+  - `"continue"` is `continue` statements.
+  - `"debugger"` is `debugger` statements.
+  - `"default"` is `default` clauses in `switch` statements.
+  - `"directive"` is directive prologues. This matches directives; e.g. `"use strict"`.
+  - `"do"` is `do-while` statements. This matches all statements that the first token is `do` keyword.
+  - `"empty"` is empty statements.
+  - `"export"` is `export` declarations.
+  - `"expression"` is expression statements.
+  - `"for"` is `for` loop families. This matches all statements that the first token is `for` keyword.
+  - `"function"` is function declarations.
+  - `"if"` is `if` statements.
+  - `"iife"` is immediately invoked function expression statements. This matches calls on a function expression, optionally prefixed with a unary operator.
+  - `"import"` is `import` declarations.
+  - `"let"` is `let` variable declarations, both single-line and multiline.
+  - `"return"` is `return` statements.
+  - `"switch"` is `switch` statements.
+  - `"throw"` is `throw` statements.
+  - `"try"` is `try` statements.
+  - `"var"` is `var` variable declarations, both single-line and multiline.
+  - `"while"` is `while` loop statements.
+  - `"with"` is `with` statements.
+  - `"type"` is `type` statements.
+  - `"interface"` is `interface` statements.
+
+- `MODIFIER_TYPE` is one of the following
+  - `"singleline-"` only treats single line `STATEMENT_TYPE`s.
+  - `"multiline-"` only treats multi line `STATEMENT_TYPE`s.
```

Comment:
as mentioned in the last review:

> Instead of copy-pasting the docs from eslint, please just list the changes.

Reply:
> Unless i'm mistaken - isn't the only change you've made to the rule changing the schema here to be unvalidated?

There is a lot more changes actually. Apart from not validating, here are a list of changes:

- Previously, `multiline-` and `singleline-` were only available for certain `keywords` such as `const`. This now applies to any statement type (e.g. `singleline-for` statements), so the documentation that I `copy-pasted` is necessary (there is a change of syntax).
- Some keywords have been changed.
- This version now uses the ESTree to validate a CommonJS import/export rather than relying on regular expressions.
- Some statement tests did not account for all cases, such as `(1, 2, () => void 0)()`, so this has been taken into account.

In general, a lot of refactoring has been done to generalize and shorten the code. Moreover, since utility functions were internal to the plugin, some functions had to be copy and pasted.

Reply:
Extension rules should make as few changes as possible.

An extension rule should not rework or add new features to the rule that are not related to adding support for typescript features (or to use type information to improve the rule).

If you want to add new features and change options - this should first occur in the ESLint repo.

Why? Because people expect to be able to just drop in extension rules as 1:1 replacements.
If they drop in an extension rule and it works completely differently - then it's going to cause significant pain and friction for users.

Additionally, the more changes we make to a fork - the harder it is to merge in upstream changes.

Reply:
Sounds reasonable. I'll send a PR in the ESLint repository.